### PR TITLE
Add date with time support in DateConvert

### PR DESCRIPTION
**Pull request by danielbadawi**

### CODE REVIEW

**Comment by felipeweb** [comment]
I think it's better to create a new convert to DateTime instead of adapting DateConverter

**Comment by danielbadawi** [comment]
DateTime from Joda Time? But if my project doesn't use it?

**Comment by felipeweb** [comment]
Sorry, Now i see your point.

2015-02-23 18:04 GMT-03:00 Daniel Badawi <notifications@github.com>:

> DateTime from Joda Time? But if my project doesn't use it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/vraptor4/pull/946#issuecomment-75631373>.
>



-- 
Felipe Oliveira
Tel: 11 99930-9974
Caelum | Ensino & Inovação
www.caelum.com.br

File: vraptor-core/src/main/java/br/com/caelum/vraptor/converter/DateConverter.java [hunk 1]
```diff
@@ -61,16 +61,18 @@ public Date convert(String value, Class<? extends Date> type) {
...
+				DateFormat formatDate = DateFormat.getDateInstance(DateFormat.SHORT, locale);  
+			try {  
+				return formatDate.parse(value);  
+			} catch (ParseException pe1) {  
+				throw new ConversionException(new ConversionMessage(INVALID_MESSAGE_KEY, value));
+			}
 		}
 	}
 
-	protected DateFormat getDateFormat() {
```
Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/converter/DateConverter.java]: we can't just remove this method, there is apps/users especializing this class and 
overriding this method to change `DateFormat`. If we drop it, we'll break them all

**Comment by Turini** [comment]
about the PR, it sounds nice but I'm not sure if we should change this behavior.
Our converter serves most cases, and if users want to change it, they can easily
specialize the default behavior. Besides that, if we change it we should change 
the `CalendarConverter` too (it means, adopting the fallback strategy as a pattern 
for all of our default converters). Lets wait for more opnions. Thoughts? `@all`

**Comment by lucascs** [comment]
we can add this extension to the docs instead, so users could easily copy and paste with their intended date format. No need for trying several formats here.

Also, trying several formats is slower, and can harm performance if there are a lot of date conversions on the request.

**Comment by danielbadawi** [comment]
I agree that trying several (actually is two) formats is slower. But Date with time is quite common. VRaptor 3 tries two different Date formats in its converter. I know that VRaptor 4 has no backwards compatibility with Vraptor 3, but IMHO ignoring the time in Date on a http post/get sounds more like a bug and only confuses the developer where the time is lost.

Of course this PR is just a suggestion. In my project I resolved this especializing the DateConverter.

**Comment by csokol** [comment]
Maybe we should let the user to define the pattern that he wants to use
with an annotation on the parameter or the field that is being converted (I
think spring does something like that).

**Comment by asouza** [comment]
I like the @DateTimeFormat annotation used in the Spring MVC. We could try
the same approach, what do you think?


Em ter, 24 de fev de 2015 19:58, csokol <notifications@github.com> escreveu:

**Comment by lucascs** [comment]
We don't have information about the conversion point on the converter =/

**Comment by csokol** [comment]
I think we could check how @Load from vraptor-hibernate works and do the
same, what do you think?

Anyway, can we change this in our API? I think this was an issue I had when
trying to create a converter of Optional<T>. I don't know if I was doing
something wrong with generics but I couldn't get a ParameterizedType to
discover the actual type of T. If vraptor provided the actual conversion
point we could do a lot more crazy things.

Em 0:23 qua, 25/02/2015, Lucas Cavalcanti <notifications@github.com>
escreveu:

We don't have information about the conversion point on the converter =/

**Comment by lucascs** [comment]
We'd have to change the converter interface to receive the conversion point...

the same logic as @Load doesn't work because the converter works for the whole object tree, not just for the parameters.

**Comment by Turini** [comment]
and do you guys think it's worth? IMO it doesn't, since we can do:

```
@Specializes @Convert(Date.class)
public class Custom DateConverter extends DateConverter {
    
    @Override
    protected DateFormat getDateFormat() {
        // any format you want
    }
}
```
Or there is more advantages using this approach that I'm not seeing?

**Comment by asouza** [comment]
The advantage is that you would not have to especialize any class. If the
thing works out of the box, at least for me, is always better.

Em sex, 27 de fev de 2015 às 11:10, Rodrigo Turini <notifications@github.com>
escreveu:

**Comment by csokol** [comment]
I hate boilerplate code :-)

Can we create a new interface that receives the conversion point? So we
wont break any user code...

**Comment by Turini** [comment]
of course, I can see it. The question is if we can take more advantage of this 
approach or this will be the only gain... not means that the gain is unimportant,
but we'd basically re-write our converter mechanism... does it worth for `4.2`?

**Comment by csokol** [comment]
I think it will be needed to write converter for Optional<T>, I've tried
some days ago unsuccessfully (it could be because I'm stupid and don't know
how to handle generics hell also)

**Comment by danielbadawi** [comment]
@Turini, overriding getDateFormat() doesn't allow multi Date format. It means, we have the same Date format in the whole application

**Comment by Turini** [comment]
@csokol its a nice case of use, I'm guessing it wont work with our current 
code (didn't try yet, but I'll asap). Anyway, can you open an issue about it? 

@danielbadawi users can override the convert method in most cases (and 
for your case, that's kinda unusual, you can override the `convert` method
adding the fallback strategy just like you did in this PR, but only for your app).

**Comment by csokol** [comment]
Should I open this issue here at caelum/vraptor4 or at caelum/vraptor-java8?

**Comment by Turini** [comment]
caelum/vraptor4 (but we can move if we discover that doesn't work specifically with optional)

**Comment by lucascs** [comment]
Adding the conversionPoint to the interface would be generally useful...
not exactly easy to implement, but would allow the OptionalConverter and
the `@DateTimeFormat` annotation.

**Comment by Turini** [comment]
Yeap, it'd be great to allow optional converter and stuff like that. :+1: 
@csokol opened the issue https://github.com/caelum/vraptor4/issues/952 about that, so I think we can close 
this PR and work on it for now. Is it ok for you @danielbadawi? thks

**Comment by danielbadawi** [comment]
@Turini it's ok!
thanks

**Comment by Turini** [comment]
thank you @danielbadawi!

**Pull request closed without merging**